sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -17,61 +17,203 @@
+  private def pushDownFilters(plan: LogicalPlan) = plan.transform {
+    // update the scan builder with filter push down and return a new plan with filter pushed
+    case filter @ Filter(_, sHolder: ScanBuilderHolder) =>
+      val (filters, _, _) = collectFilters(filter).get
+
+      val normalizedFilters =
+        DataSourceStrategy.normalizeExprs(filters, sHolder.relation.output)
       val (normalizedFiltersWithSubquery, normalizedFiltersWithoutSubquery) =
         normalizedFilters.partition(SubqueryExpression.hasSubquery)
 
       // `pushedFilters` will be pushed down and evaluated in the underlying data sources.
       // `postScanFilters` need to be evaluated after the scan.
       // `postScanFilters` and `pushedFilters` can overlap, e.g. the parquet row group filter.
       val (pushedFilters, postScanFiltersWithoutSubquery) = PushDownUtils.pushFilters(
-        scanBuilder, normalizedFiltersWithoutSubquery)
+        sHolder.builder, normalizedFiltersWithoutSubquery)
       val postScanFilters = postScanFiltersWithoutSubquery ++ normalizedFiltersWithSubquery
+      val output = sHolder.output
+
+      logInfo(
+        s"""
+           |Pushing operators to ${sHolder.relation.asInstanceOf[DataSourceV2Relation].name}
+           |Pushed Filters: ${pushedFilters.mkString(", ")}
+           |Post-Scan Filters: ${postScanFilters.mkString(",")}
+         """.stripMargin)
 
+      val projectionOverSchema = ProjectionOverSchema(output.toStructType)
+      val projectionFunc = (expr: Expression) => expr transformDown {
+        case projectionOverSchema(newExpr) => newExpr
+      }
+
+      val filterCondition = postScanFilters.reduceLeftOption(And)
+      val newFilterCondition = filterCondition.map(projectionFunc)
+      newFilterCondition.map(Filter(_, sHolder)).getOrElse(sHolder)
+  }
+
+  def pushdownAggregate(plan: LogicalPlan): LogicalPlan = plan.transform {
+    // update the scan builder with agg pushdown and return a new plan with agg pushed
+    case aggNode @ Aggregate(groupingExpressions, resultExpressions, child) =>
+      child match {
+        case ScanOperation(project, filters, sHolder: ScanBuilderHolder) =>
+          sHolder.builder match {
+            case r: SupportsPushDownAggregates =>
+              if (filters.isEmpty || r.supportsPushDownAggregateWithFilter()) {
+                if (r.supportsGlobalAggregatePushDownOnly() && groupingExpressions.nonEmpty) {
+                  aggNode // return original plan node
+                } else {
+                  val aggregates = getAggregateExpression(resultExpressions, project, sHolder)
+                  val pushedAggregates = PushDownUtils
+                    .pushAggregates(sHolder.builder, aggregates, groupingExpressions)
+                  logInfo(
+                    s"""
+                       |Pushed Aggregates: ${pushedAggregates.aggregateExpressions.mkString(", ")}
+                    """.stripMargin)
+                  aggNode
```

Comment:
Since agg pushdown and column pruning is mutually exclusive, we can stop here, rewrite the plan, and create `DataSourceV2ScanRelation`

Reply:
A bit more thinking: if the query is `SELECT a, max(c) FROM t GROUP BY a, b`, then column pruning still applies as we can tell the source to not output column `b`.

This is perhaps not worthwhile, as if a source can support agg pushdown, it should be cheap to output all the grouping columns. So after agg pushdown, I think it's fine for the source to always output the grouping columns and the aggregate results.

Reply:
Changed.